lib/syntax_tree/visitor/environment.rb

```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+module SyntaxTree
+  # The environment class is used to keep track of local variables and arguments
+  # inside a particular scope
+  class Environment
+    # [Array[Local]] The local variables and arguments defined in this
+    # environment
+    attr_reader :locals
+
+    # This class tracks the occurrences of a local variable or argument
+    class Local
+      # [Symbol] The type of the local (e.g. :argument, :variable)
+      attr_reader :type
+
+      # [Array[Location]] The locations of all occurrences of this local,
+      # including its definition
```

Comment:
Does it make sense to track the definition of a local separately, so that tools can use that information? For example, that would allow Ruby LSP to implement `Go to Definition` for local variables.

Reply:
I think it makes sense. It would always be the first location in the array. Do you think we should store it separately or just provide a method that returns the first location? Something like

```ruby
def definition_location
  @locations.first
end
```

Reply:
Unfortunately it wouldn't always be the first necessarily. 

```ruby
a = 1 if a
```

In this case the var_ref is found before the var_field.

I think I'd definitely like to track the definition and usages separately. They're different types usually anyway, since all of the definitions will be `*Field` nodes, and the usages will be `VarRef` nodes.

Reply:
That's fair. What would you expect for locals that are assigned multiple times or conditionally? E.g.:
```ruby
# Which one of these are stored as the definition location? Both?
if something
  a = 1
else
  a = 2
end
# ...

# How about this one? Do we also include it as a definition location?
a = 5
```

Reply:
Oof that's hard. I would say it would be an array of the first and second, but that's going to get into some pretty in-depth flow analysis that maybe we shouldn't do. How about just separating the assignments from the references?

Reply:
I think that makes sense, especially as a first iteration. Maybe in the future we can take a stab at more complex analysis.

The one thing that can be slightly odd is for arguments. Should we include their definitions together with re-assignments? It might be fine as long as we document it and name the array something like `definitions` and not `assignments`.
```ruby
def foo(a) # This would be a definition
  something(a) # This would be a usage
  a = 123 # This would be another "definition"
end
```

Reply:
Yeah `definitions` makes a lot of sense.

Reply:
I split usages and definitions in the latest commit.

I also refactored the tests a little bit to check what is saved in the environment, rather than saving the nodes. This uncovered a couple of bugs that I fixed.

The one thing that it uncovered and I'm not sure what we could do is that rest variables in pinning are considered `VCall`, not a `VarRef`.

I think we had spoken about this a while back and you mentioned this comes from Ripper itself, so we may just have to leave it as is for now.

```ruby
def foo
  case [1, 2]
  in ^a, *rest # => this is a var_field and gets put in definitions
    puts rest # => this is considered a vcall and is ignored
  end
end
```